docs/developer/logging-guidelines.md

```diff
@@ -0,0 +1,407 @@
+   ```
+
+   > An error message includes the stack trace by default.
+4. A message with a Kubernetes resource:
+
+   ```go
+   // hr implements client.Object (controller-runtime)
+   // hr is *v1beta1.HTTPRoute
+    logger.Info("Orocessed resource",
+        "resource", hr)
+   ```
+
+   ```json
+   {"level":"info","ts":"2023-07-20T15:10:03-04:00","msg":"Processed resource","resource":{"apiVersion":"gateway.networking.k8s.io/v1beta1","kind":"HTTPRoute","namespace":"test","name":"hr-1"}}
+   ```
+
+   > The resource must include `TypeMeta`, otherwise its `apiVersion` and `kind` will not be printed.
+5. A message with a component name:
+
+   ```go
+    logger = logger.WithName("component-a")
+    logger.Info("Hello")
+   ```
+
+   ```json
+    {"level":"info","ts":"2023-07-20T15:10:03-04:00","logger":"component-a","msg":"Hello"}
+   ```
+
+6. A message with extra context:
+
+   ```go
+    logger.Info("info message",
+        "status", "ok",
+        "generation", 123)
+   ```
+
+   ```json
+   {"level":"info","ts":"2023-07-20T15:10:03-04:00","msg":"Info message","status":"ok","generation":123}
+   ```
+
+7. An info message with a stack trace configured for the `info` level:
+
+   ```go
+    logger.Info("Info message")
+   ```
+
+   ```json
+   {"level":"info","ts":"2023-07-20T15:10:03-04:00","msg":"Info message","stacktrace":"main.main\n\t/<REDACTED for the document>.go:100\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"}
+   ```
+
+#### Message Guidelines
+
+1. Start a message with an uppercase letter. For example, `"Reloaded nginx"`.
+2. Do not put dynamic values in the message, add them as key values (extra context):
+
+   DO NOT:
+
+   ```go
+    logger.Info(fmt.Sprintf("Got response with status %q", "ok"))
+   ```
+
+   DO:
+
+   ```go
+    logger.Info("Got response",
+        "status", "ok")
+   ```
+
+3. For readability, put each key-value pair on a separate line, as in the previous example.
+4. Use camelCase for key names.
+5. When dealing with a Kubernetes resource, use `resource` as a key:
+
+   ```go
+    logger.Info("processed resource",
+        "resource", hr)
+   ```
+
+6. Don't put structs as values expect for Kubernetes resources, to prevent messy and potentially unbounded output.
+
+### Log Formatting
+
+By default, log messages are formatted as JSON, so that external systems can easily parse them.
+
+For the developer, log messages are formatted as text strings (except key/values), for human-friendly parsing. For
+example:
+
+```text
+2023-07-21T12:41:37.640-0400    INFO    processed resource      {"resource": {"apiVersion": "gateway.networking.k8s.io/v1beta1", "kind": "HTTPRoute", "namespace": "test", "name": "hr-1"}}
+```
+
+The formatting is controlled during the logger initialization.
+
+### When to Log
+
+Below are the cases for when to log a message:
```

Comment:
Is this intended to be a complete list? Or just some examples?

Reply:
each, the complete list at high level. is there anything important missing?

Reply:
I just wonder if we are being too strict and as the project develops these cases may evolve as well. Maybe we can call these examples instead of cases? "Below are some examples of when to log a message"